[FIX] mrp : wrong expected duration in WC with capacity

nmarotte opened the pull request:
Issue: In work centers with a capacity different than 1, the expected duration was wrongly computed

Steps to reproduce :
 1) Install Manufacture, enable Work Centers in Settings
 2) Create a Work Center, Capacity = 2
 3) Create a Product, with Manufacture as route
 4) Create a Bill of Material :
   Add any component
   Add a operation line :
     Work Center = created at 2)
     Duration Computation = based on tracked time last 1 WO
 5) Create a Manufacturing Order for the product 3), confirm
 6) Set the real duration to 00:10, mark as done
 7) Check the BoM Structure & Cost of the BoM 4)
 -> Operations is 00:20

Why is that a bug:
 To compute the expected duration, in mrp_workorder.py L691 we already take into consideration the capacity of the work center.
 When computing the time_cycle, we must give a time_cycle for 1 product made with capacity 1, so we must first normalize the qty_produced fetch from the DB to make like if we only produced the quantity once but in the same amount of time

opw-2562983

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/75078).

amoyaux commented:
Isn't it possible to use array_agg instead of removing read_group?

nmarotte commented:
> Isn't it possible to use array_agg instead of removing read_group?

@amoyaux I'm not really sure, I asked my team but they don't think it could be use for that case
Since the limit keyword is applied after the group by, the limit is currently not working (have a compute on last 1 duration, create 2 MO and then print in the code, you'll see the quantity is the sum of the 2 quantities instead of just 1 quantity) so we don't really see how to use the array_agg for that purpose :confused:

amoyaux commented:
@nmarotte Okay but then also add the explanation of the limit limiting the number of groups and not the number of records in the group. Also it would be great to add a test covering this basic usecase.

nmarotte commented:
@amoyaux It's done, I've added the test, it indeed fails (at the assertNotEqual) before my changes and then succeed after my commit

amoyaux commented:
Imagine the usecase:

WC capacity = 2
WO 1 produce 1 -> 10s
WO 2 produce 2 -> 10s

Total produced 3. What should be the time cycle for you?

nmarotte commented:
> Imagine the usecase:
> 
> WC capacity = 2
> WO 1 produce 1 -> 10s
> WO 2 produce 2 -> 10s
> 
> Total produced 3. What should be the time cycle for you?

I see what you mean @amoyaux, I changed it again and added more tests to cover that, tell me what you think :smile:

amoyaux commented on addons/mrp/models/mrp_routing.py:
```diff
@@ -49,14 +50,24 @@ def _compute_time_cycle(self):
+            total_qty_per_capacity_for_duration = 0  # Never 0 unless infinite item['workcenter_id'].capacity
+            for item in data:
+                total_duration += item['duration']
+                total_qty_per_capacity_for_duration += math.ceil(item['qty_produced'] / item['workcenter_id'].capacity or 1.0)
```
You could use float_round with rounding_method='DOWN' to use odoo tools and avoid a bugifx later on a strange rounding issue

nmarotte replied:
@amoyaux Don't you mean rounding_method='UP' ? Since we are using math.ceil (1.001 becomes 2)
It is also used here https://github.com/odoo-dev/odoo/blob/11c02365d52c973b6c58685cbe52cf3ea94c5fd0/addons/mrp/models/mrp_workorder.py#L691 so I changed my variable names too to line up with the code to see more easily the link between the two (divide by average cycle_number in mrp_routing.py and multiply by workorder cycle_number in mrp_workorder.py

amoyaux commented:
Except my small comment, it seems great

amoyaux commented:
robodoo r+
